pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CoalesceTransformFunction.java

```diff
@@ -0,0 +1,361 @@
+      boolean hasNonNullValue = false;
+      for (int j = 0; j < width; j++) {
+        // Consider value as null only when null option is enabled.
+        if (nullBitMaps[j] != null && nullBitMaps[j].contains(i)) {
+          continue;
+        }
+        if (!filledData.contains(j)) {
+          filledData.add(j);
+          data[j] = _transformFunctions[j].transformToBigDecimalValuesSV(projectionBlock);
+        }
+        hasNonNullValue = true;
+        results[i] = data[j][i];
+        break;
+      }
+      if (!hasNonNullValue) {
+        results[i] = (BigDecimal) NullValueUtils.getDefaultNullValue(_dataType);
+      }
+    }
+    return results;
+  }
+
+
+  /**
+   * Get transform String results based on store type.
+   * @param projectionBlock
+   */
+  private String[] getStringTransformResults(ProjectionBlock projectionBlock) {
+    int length = projectionBlock.getNumDocs();
+    String[] results = new String[length];
+    int width = _transformFunctions.length;
+    RoaringBitmap[] nullBitMaps = getNullBitMaps(projectionBlock, _transformFunctions);
+    String[][] data = new String[width][length];
+    RoaringBitmap filledData = new RoaringBitmap(); // indicates whether certain column has be filled in data.
+    for (int i = 0; i < length; i++) {
+      boolean hasNonNullValue = false;
+      for (int j = 0; j < width; j++) {
+        // Consider value as null only when null option is enabled.
+        if (nullBitMaps[j] != null && nullBitMaps[j].contains(i)) {
+          continue;
+        }
+        if (!filledData.contains(j)) {
+          filledData.add(j);
+          data[j] = _transformFunctions[j].transformToStringValuesSV(projectionBlock);
+        }
+        hasNonNullValue = true;
+        results[i] = data[j][i];
+        break;
+      }
+      if (!hasNonNullValue) {
+        results[i] = (String) NullValueUtils.getDefaultNullValue(_dataType);
+      }
+    }
+    return results;
+  }
+
+  @Override
+  public String getName() {
+    return TransformFunctionType.COALESCE.getName();
+  }
+
+  @Override
+  public void init(List<TransformFunction> arguments, Map<String, DataSource> dataSourceMap) {
+    int argSize = arguments.size();
+    Preconditions.checkArgument(argSize > 0, "COALESCE needs to have at least one argument.");
+    _transformFunctions = new TransformFunction[argSize];
+    for (int i = 0; i < argSize; i++) {
+      TransformFunction func = arguments.get(i);
+      Preconditions.checkArgument(func instanceof IdentifierTransformFunction,
+          "Only column names are supported in COALESCE.");
+      FieldSpec.DataType dataType = func.getResultMetadata().getDataType().getStoredType();
+      if (_dataType == null) {
+        _dataType = dataType;
+      } else {
+        Preconditions.checkArgument(dataType.equals(_dataType), "Argument types have to be the same.");
+      }
+      _transformFunctions[i] = func;
+    }
+  }
+
+  @Override
+  public TransformResultMetadata getResultMetadata() {
```

Comment:
(minor) Let's calculate the result metadata in the `init()` and store it in a member variable. It has 2 benefits:
1. Fail fast when the type cannot be supported
2. Prevent calculating result metadata multiple times